File: zig/src/collectors/docker.zig
const std = @import("std");
const utils = @import("../utils.zig");
const report = @import("../models/report.zig");

/// Collect Docker information
pub fn collectDockerInfo(allocator: std.mem.Allocator) ?report.DockerInfo {
    // Check if docker command exists
    if (!utils.commandExists("docker")) {
        return report.DockerInfo{
            .available = false,
        };
    }

    // Check if Docker daemon is running
    _ = utils.runCommand(allocator, &[_][]const u8{ "docker", "info" }) catch {
        return report.DockerInfo{
            .available = true,
            .daemon_running = false,
        };
    };

    var containers = std.ArrayList(report.ContainerInfo).empty;

    // Get container list
    const ps_output = utils.runCommand(allocator, &[_][]const u8{
        "docker", "ps", "-a", "--format", "{{.Names}}\t{{.Image}}\t{{.Status}}\t{{.State}}\t{{.CreatedAt}}",
    }) catch "";

    if (ps_output.len > 0) {
        var lines = std.mem.splitScalar(u8, ps_output, '\n');
        while (lines.next()) |line| {
            if (line.len == 0) continue;
            if (parseContainerLine(allocator, line)) |container| {
                containers.append(allocator, container) catch continue;
            }
        }
    }

    // Count running/stopped containers
    var running: i32 = 0;
    var stopped: i32 = 0;
    for (containers.items) |c| {
        if (std.mem.eql(u8, c.state, "running")) {
            running += 1;
        } else {
            stopped += 1;
        }
    }

    // Get image count
    const images_output = utils.runCommand(allocator, &[_][]const u8{
        "docker", "images", "--format", "{{.Size}}",
    }) catch "";

    var image_count: i32 = 0;
    var total_image_size: i64 = 0;
    if (images_output.len > 0) {
        var img_lines = std.mem.splitScalar(u8, images_output, '\n');
        while (img_lines.next()) |l| {
            if (l.len > 0) {
                image_count += 1;
                total_image_size += @intCast(parseSizeString(l));
            }
        }
    }

    // Get dangling images size
    var dangling_size: i64 = 0;
    const dangling_output = utils.runCommand(allocator, &[_][]const u8{
        "docker", "images", "-f", "dangling=true", "--format", "{{.Size}}",
    }) catch "";

    if (dangling_output.len > 0) {
        var lines = std.mem.splitScalar(u8, dangling_output, '\n');
        while (lines.next()) |l| {
            if (l.len > 0) {
                dangling_size += @intCast(parseSizeString(l));
            }
        }
    }

    const container_slice: ?[]const report.ContainerInfo = if (containers.items.len == 0)
        null
    else
        containers.toOwnedSlice(allocator) catch null;

    return report.DockerInfo{
        .available = true,
        .daemon_running = true,
        .containers = container_slice,
        .running_count = running,
        .stopped_count = stopped,
        .image_count = image_count,
        .total_image_size = total_image_size,
        .dangling_images_size = dangling_size,
    };
}

fn parseContainerLine(allocator: std.mem.Allocator, line: []const u8) ?report.ContainerInfo {
    var parts = std.mem.splitScalar(u8, line, '\t');

    const name = parts.next() orelse return null;
    const image = parts.next() orelse return null;
    const status = parts.next() orelse return null;
    const state = parts.next() orelse return null;
    const created = parts.next() orelse "";

    return report.ContainerInfo{
        .name = allocator.dupe(u8, name) catch return null,
        .image = allocator.dupe(u8, image) catch return null,
        .status = allocator.dupe(u8, status) catch return null,
        .state = allocator.dupe(u8, state) catch return null,
        .created = allocator.dupe(u8, created) catch "",
    };
}

fn parseSizeString(s: []const u8) u64 {
    const trimmed = utils.trim(s);
    if (trimmed.len == 0) return 0;

    var num_end: usize = 0;
    for (trimmed) |c| {
        if ((c >= '0' and c <= '9') or c == '.') {
            num_end += 1;
        } else {
            break;
        }
    }

    if (num_end == 0) return 0;

    const num = utils.parseFloat(trimmed[0..num_end], 0);
    const unit = if (num_end < trimmed.len) trimmed[num_end..] else "";

    const multiplier: f64 = if (std.mem.startsWith(u8, unit, "TB") or std.mem.startsWith(u8, unit, "T"))
        1024 * 1024 * 1024 * 1024
    else if (std.mem.startsWith(u8, unit, "GB") or std.mem.startsWith(u8, unit, "G"))
        1024 * 1024 * 1024
    else if (std.mem.startsWith(u8, unit, "MB") or std.mem.startsWith(u8, unit, "M"))
        1024 * 1024
    else if (std.mem.startsWith(u8, unit, "KB") or std.mem.startsWith(u8, unit, "K") or std.mem.startsWith(u8, unit, "kB"))
        1024
    else
        1;

    return @intFromFloat(num * multiplier);
}

File: go/collectors/docker.go
package collectors

import (
	"encoding/json"
	"os/exec"
	"strconv"
	"strings"

	"machinestate/models"
)

// dockerContainer represents the JSON output from docker ps
type dockerContainer struct {
	Names  string `json:"Names"`
	Image  string `json:"Image"`
	Status string `json:"Status"`
	State  string `json:"State"`
	// RunningFor is available in docker ps --format
	CreatedAt string `json:"CreatedAt"`
}

// dockerImage represents the JSON output from docker images
type dockerImage struct {
	Size string `json:"Size"`
}

// dockerDiskUsage represents the JSON output from docker system df
type dockerDFOutput struct {
	Type        string `json:"Type"`
	TotalCount  int    `json:"TotalCount"`
	Size        string `json:"Size"`
	Reclaimable string `json:"Reclaimable"`
}

// CollectDockerInfo gathers Docker container and image information
func CollectDockerInfo() models.DockerInfo {
	info := models.DockerInfo{
		Containers: []models.ContainerInfo{},
	}

	// Check if docker is available
	if _, err := exec.LookPath("docker"); err != nil {
		return info
	}
	info.Available = true

	// Check if daemon is running
	cmd := exec.Command("docker", "info")
	if err := cmd.Run(); err != nil {
		info.Available = true
		info.DaemonRunning = false
		return info
	}
	info.DaemonRunning = true

	// Get running containers
	info.Containers, info.RunningCount = getContainers("running")

	// Get stopped containers count
	_, info.StoppedCount = getContainers("exited")

	// Get images info
	info.ImageCount, info.TotalImageSize = getImagesInfo()

	// Get dangling images size
	info.DanglingImages = getDanglingImagesSize()

	return info
}

func getContainers(status string) ([]models.ContainerInfo, int) {
	var containers []models.ContainerInfo

	cmd := exec.Command("docker", "ps", "-a",
		"--filter", "status="+status,
		"--format", "{{json .}}")
	output, err := cmd.Output()
	if err != nil {
		return containers, 0
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	for _, line := range lines {
		if line == "" {
			continue
		}

		var dc dockerContainer
		if err := json.Unmarshal([]byte(line), &dc); err != nil {
			continue
		}

		container := models.ContainerInfo{
			Name:    strings.TrimPrefix(dc.Names, "/"),
			Image:   dc.Image,
			Status:  dc.Status,
			State:   dc.State,
			Created: dc.CreatedAt,
		}

		// Get container stats for CPU/Memory (only for running containers)
		if status == "running" {
			container.CPUPercent, container.MemoryPercent = getContainerStats(container.Name)
		}

		containers = append(containers, container)
	}

	return containers, len(containers)
}

func getContainerStats(name string) (float64, float64) {
	cmd := exec.Command("docker", "stats", name, "--no-stream", "--format", "{{.CPUPerc}},{{.MemPerc}}")
	output, err := cmd.Output()
	if err != nil {
		return 0, 0
	}

	parts := strings.Split(strings.TrimSpace(string(output)), ",")
	if len(parts) != 2 {
		return 0, 0
	}

	cpuStr := strings.TrimSuffix(parts[0], "%")
	memStr := strings.TrimSuffix(parts[1], "%")

	cpu, _ := strconv.ParseFloat(cpuStr, 64)
	mem, _ := strconv.ParseFloat(memStr, 64)

	return cpu, mem
}

func getImagesInfo() (int, int64) {
	cmd := exec.Command("docker", "images", "--format", "{{.Size}}")
	output, err := cmd.Output()
	if err != nil {
		return 0, 0
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	count := 0
	var totalSize int64 = 0

	for _, line := range lines {
		if line == "" {
			continue
		}
		count++
		totalSize += parseSize(line)
	}

	return count, totalSize
}

func getDanglingImagesSize() int64 {
	cmd := exec.Command("docker", "images", "-f", "dangling=true", "--format", "{{.Size}}")
	output, err := cmd.Output()
	if err != nil {
		return 0
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	var totalSize int64 = 0

	for _, line := range lines {
		if line == "" {
			continue
		}
		totalSize += parseSize(line)
	}

	return totalSize
}

// parseSize converts Docker size strings like "1.23GB" to bytes
func parseSize(sizeStr string) int64 {
	sizeStr = strings.TrimSpace(sizeStr)
	if sizeStr == "" {
		return 0
	}

	var multiplier int64 = 1
	sizeStr = strings.ToUpper(sizeStr)

	if strings.HasSuffix(sizeStr, "GB") {
		multiplier = 1024 * 1024 * 1024
		sizeStr = strings.TrimSuffix(sizeStr, "GB")
	} else if strings.HasSuffix(sizeStr, "MB") {
		multiplier = 1024 * 1024
		sizeStr = strings.TrimSuffix(sizeStr, "MB")
	} else if strings.HasSuffix(sizeStr, "KB") {
		multiplier = 1024
		sizeStr = strings.TrimSuffix(sizeStr, "KB")
	} else if strings.HasSuffix(sizeStr, "B") {
		sizeStr = strings.TrimSuffix(sizeStr, "B")
	}

	value, err := strconv.ParseFloat(sizeStr, 64)
	if err != nil {
		return 0
	}

	return int64(value * float64(multiplier))
}
